<a href="https://colab.research.google.com/github/sai1519/My-projects-and-programs/blob/Machine-Learning/Saishta's_Assignment_8_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Program 1

### We shall use the same dataset used in previous assignment - digits. Make a 80- 20 train/test split.
### [Hint: Explore datasets module from scikit learn]

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

# Loading the dataset.
digit = load_digits()

# Choosing x and y data.
x = digit.data
y = digit.target

# Splitting the data.
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=42)

In [ ]:
x_train.shape

(1437, 64)

In [ ]:
y_train.shape

(1437,)

In [ ]:
x_test.shape

(360, 64)

In [ ]:
y_test.shape

(360,)

# Program 2
### Using scikit learn perform a LDA on the dataset. Find out the number of components in the projected subspace.
### [Hint: Refer to discriminant analysis module of scikit learn]

In [ ]:
# Performing LDA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda = LinearDiscriminantAnalysis()
lda.fit(x_train,y_train)

LinearDiscriminantAnalysis()

In [ ]:
# Getting the explained variance ration for each component.
evr = lda.explained_variance_ratio_

# Calculating the cumulative explained variance.
cev = evr.cumsum()

# Finding the number of components.
dvt = 0.95
ncomp = np.argmax(cev >= dvt) +1
ncomp

8

# Program 3
### Transform the dataset and fit a logistic regression and observe the accuracy. Compare it with the previous model based on PCA in terms of accuracy and model complexity.
### [Hint: Project both the train and test samples to the new subspace]

In [ ]:
# Transforming the data.
new_x_train=lda.transform(x_train)
new_x_test=lda.transform(x_test)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr=LogisticRegression()
lr.fit(new_x_train,y_train)

LogisticRegression()

In [ ]:
y_pred = lr.predict(new_x_test)

In [ ]:
accuracy_lda =accuracy_score(y_test,y_pred)
accuracy_lda

0.9472222222222222

In PCA model the accuracy was 95% and here its 94.7% which is almost same or slightly less than PCA.